src/rusty/src/lib.rs

```diff
@@ -0,0 +1,64 @@
+mod bridge;
+use bridge::*;
+mod dns_headers;
+
+use std::alloc::{GlobalAlloc, Layout, System};
+use std::ptr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::{Duration, Instant};
+
+// We keep track of all memory allocated by Rust code, refusing new allocations if it exceeds
+// 128MB.
+//
+// Note that while Rust's std, in general, should panic in response to a null allocation, it
+// is totally conceivable that some code will instead dereference this null pointer, which
+// would violate our guarantees that Rust modules should never crash the entire application.
+//
+// In the future, as upstream Rust explores a safer allocation API (eg the Alloc API which
+// returns Results instead of raw pointers, or redefining the GlobalAlloc API to allow
+// panic!()s inside of alloc calls), we should switch to those, however these APIs are
+// currently unstable.
+const TOTAL_MEM_LIMIT_BYTES: usize = 128 * 1024 * 1024;
```

Comment:
> refusing new allocations if it exceeds 128MB.

So when Rust tries to allocate more than 128mb bitcoind will crash? I dropped the limit to something tiny and `bitcoind` would crash on startup.

Reply:
The reason it crashed all of bitcoind when this is tiny is because this allocation is outside out `catch_unwind` https://github.com/bitcoin/bitcoin/pull/16834/files#diff-57a33003fdd95417a0d7d83201893c62R69

Reply:
Indeed, I think its fine, but just to make your testing easier, I went ahead and wrapped the whole function in a unwind.